### ISSUE TITLE
How to use automl with non-numerical data

### ISSUE THREAD

**Issue by lesshaste** [automl/auto-sklearn#121]
I have a dataset where  some of the fields are strings such as "happy" or "sad".  In order to use automl should I 

a) One-hot encode them or
b) just convert them  to integers and let automl worry about what to  do?

 In other words, does automl already have a builtin way to handle such categorical data?

**Comment by mfeurer** [comment]
Hi,
Thanks for pointing out that this is not covered in the documentation. You should convert them to integers and provide a list of strings, in which each entry describes the type of one feature. Each entry in that list must be either 'numerical' or 'categorical'. Hope that helps.

I leave this open to remind us that this should be mentioned in the docs.

**Comment by Motorrat** [comment]
@mfeurer is factorizing a preferred solution to binarizing if the column does contain more than two categories? 
As an example why that would matter check this discussion http://stackoverflow.com/questions/12569977/python-large-iterations-number-fail 
> A word of warning though: this implies that "low" safety and "high" safety are the same distance from "med" safety. You might be better off using pd.get_dummies

pd.get_dummies is fast and scales well BTW. I have tested with 5MM rows and many (tens) of categories in many columns.

**Comment by Motorrat** [comment]
@lesshaste @mfeurer the usage is actually covered in the API doc 
http://automl.github.io/auto-sklearn/stable/api.html#api
> feat_type : list, optional (default=None)
> 
>     List of str of len(X.shape[1]) describing the attribute type. Possible types are Categorical and Numerical. Categorical attributes will be automatically One-Hot encoded.

**Comment by mfeurer** [comment]
@Motorrat I'm not sure what you mean by factorization, can you please provide a link/explanation? Regarding the rather slow implementation. We're rebuilding parts of auto-sklearn to use only scikit-learn features. This should make the OneHotEncoding faster.

**Comment by Motorrat** [comment]
@mfeurer 
factorisation is this
> You should convert them to integers
I am using pandas - so this is how it goes there
 dataframe[columnx]=dataframe[columnx].astype('category').cat.codes

**Comment by mfeurer** [comment]
Passing data as integers to auto-sklearn's has the advantage that the OneHotEncoder has a configurable feature which can reduce the number of categories for a column. The OneHotEncoder also takes care that new categories are ignored during prediction time (given that they have a different integer than the values used for training).

Doing the binarization outside might also not hurt if you make sure that only the columns which were known at training time are present at test time. Also, you need to take care of imputation yourself. I must admit that I never tried this.

**Comment by Motorrat** [comment]
this pandas chain works really well, btw pd.get_dummies(dataframe.fillna(dataframe.mode().iloc[0]))

**Comment by mfeurer** [comment]
I just updated the docstring of `fit()` to make this clearer. Furthermore, I opened #274 about adding an example.

**Comment by kaiweiang** [comment]
> Do you mean you have tabular data in which a text column contains is a null (i.e. None)? Did you try using the empty string as the input to the LabelEncoder? You could then afterwards replace the transformed value by a np.NaN.

That's actually brilliant. 
Now I've done this, I'm just wondering how'd you treat or handle the dummy null/nan columns produced by the onehotencoder during the data preprocessing in the auto-sklearn?

**Comment by mfeurer** [comment]
The one hot encoder does not produce null/nan values, instead it does it bit more than the name suggests. If there are nan values in the data, the one hot encoder creates a new category 'missing'. I hope this answers the question.
